app/main.py

```diff
@@ -1,34 +1,123 @@
+            if all(not d.is_alive for d in self.decks):
+                self.is_drowned = True
+            return True
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self,
+                 ships: list[tuple[tuple[int, int], tuple[int, int]]]
+                 ) -> None:
+        self.size = 10
+        self.field = [["~" for _ in range(self.size)]
+                      for _ in range(self.size)]
+        self.ships = [Ship(*ship) for ship in ships]
+        self._place_ships()
+
+    @staticmethod
+    def _validate_field(ships: list) -> None:
+        if len(ships) != 10:
+            raise ValueError("There must be 10 ships.")
+        for ship in ships:
+            other_ships = [
+                end for temp in ships if temp != ship for end in temp
+            ]
+            for end in ship:
+                neighbours = [
+                    (end[0] - 1, end[1] - 1),
+                    (end[0] - 1, end[1]),
+                    (end[0] - 1, end[1] + 1),
+                    (end[0], end[1] - 1),
+                    (end[0], end[1] + 1),
+                    (end[0] + 1, end[1] - 1),
+                    (end[0] + 1, end[1]),
+                    (end[0] + 1, end[1] + 1),
+                ]
+
+                for neighbour in neighbours:
+                    if neighbour in other_ships:
+                        raise ValueError("The ships must not be adjacent.")
+
+        deck_size = []
+        for ship in ships:
+            deck_size.append(
+                (ship[1][0] - ship[0][0]) + (ship[1][1] - ship[0][1]) + 1
+            )
+        if sorted(deck_size) != [1, 1, 1, 1, 2, 2, 2, 3, 3, 4]:
+            raise ValueError("Wrong ship size configuration.")
+
+    def _place_ships(self) -> None:
+        for ship in self.ships:
+            for deck in ship.decks:
+                self.field[deck.row][deck.column] = u"\u25A1"
+
+    def fire(self, location: tuple[int, int]) -> str:
+        row, col = location
+        if self.field[row][col] == "~" or self.field[row][col] == ".":
+            self.field[row][col] = "."
+        else:
+            for ship in self.ships:
+                if ship.fire(row, col):
+                    self.field[row][col] = "*"
+                    if ship.is_drowned:
+                        for deck in ship.decks:
+                            self.field[deck.row][deck.column] = "x"
+                        return "Sunk!"
+                    return "Hit!"
+        return "Miss!"
```

Comment:
The `fire` method in the `Battleship` class should handle missed shots correctly. Ensure that the field is updated to indicate a miss (e.g., with a ".") and return the appropriate result when no ship is hit. This is crucial for the game logic to function as intended.